src/StoragelessSession/Http/SessionMiddleware.php

```diff
@@ -217,18 +218,32 @@ public function extractSessionContainer(Token $token = null) : Data
     /**
      * @param Data     $sessionContainer
      * @param Response $response
+     * @param Token    $originalToken
      *
      * @return Response
      *
      * @throws \InvalidArgumentException
      */
-    private function appendToken(Data $sessionContainer, Response $response) : Response
+    private function appendToken(Data $sessionContainer, Response $response, Token $originalToken = null) : Response
     {
-        if ($sessionContainer->isEmpty()) {
+        $empty = $sessionContainer->isEmpty();
+
+        if ($empty && null === $originalToken) {
```

Comment:
> we actually need to inform the client that it is now expired and shouldn't be sent again. Right?

Yup

 > I think that relying on the parsed, validated and verified token is better because we would be certain that we're not using garbage (that could be stored in the cookie and mess things up).

The idea is that we shouldn't send a `Set-Cookie` (even for forceful expiration) unless it is strictly needed.
The cookie should be *set* only when there is data.
The cookie should be *expired* only when a cookie was sent and there is no data.

All other responses should not contain the `Set-Cookie` header